File: app/src/main/java/com/bsstokes/bsdiy/api/DiyApi.java
package com.bsstokes.bsdiy.api;

import android.net.Uri;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.text.TextUtils;

import java.util.List;

import retrofit2.Response;
import retrofit2.http.GET;
import retrofit2.http.Path;
import rx.Observable;

public interface DiyApi {
    String ENDPOINT = "https://api.diy.org/";

    @GET("/skills?limit=1000&offset=0")
    Observable<Response<DiyResponse<List<Skill>>>> getSkills();

    @GET("/skills/{skill_url}/challenges")
    Observable<Response<DiyResponse<List<Challenge>>>> getChallenges(@Path("skill_url") String skillUrl);

    class Helper {
        @NonNull
        public static String normalizeUrl(@NonNull String url) {
            final Uri uri = Uri.parse(url);
            if (TextUtils.isEmpty(uri.getScheme())) {
                return uri.buildUpon().scheme("https").build().toString();
            } else {
                return uri.toString();
            }
        }
    }

    class DiyResponse<T> {
        Head head = new Head();
        public T response;

        static class Head {
            int code;
            String status = "";

            Collection collection = new Collection();

            static class Collection {
                int limit;
                int offset;

                @Override
                public String toString() {
                    return "Collection{" +
                            "limit=" + limit +
                            ", offset=" + offset +
                            '}';
                }
            }

            @Override
            public String toString() {
                return "Head{" +
                        "code=" + code +
                        ", status='" + status + '\'' +
                        ", collection=" + collection +
                        '}';
            }
        }

        @Override
        public String toString() {
            return "DiyResponse{" +
                    "head=" + head +
                    ", response=" + response +
                    '}';
        }
    }

    class Skill {
        public long id;
        // "sku":314889437,
        // "stamp":"2014-05-06T22:53:11.000Z",
        public boolean active = true;
        // "activeAt":null,
        public @Nullable String url = "";
        public @Nullable String title = "";
        public @Nullable String description = "";
        // "position": [0, 0],
        public @Nullable Icons icons = new Icons();
        public @Nullable Images images = new Images("", "", "");
        // "grammar": {
        //   "singular":"actor",
        //   "plural":"Actor",
        //   "subject":"Actor",
        //   "article":"a"
        // },
        // "pole": "art",
        public @Nullable String color;
        // "notes": ""

        public static class Icons {
            public @Nullable String small;
            public @Nullable String medium;
        }

        public static class Images {
            public Images(@Nullable String small, @Nullable String medium, @Nullable String large) {
                this.small = small;
                this.medium = medium;
                this.large = large;
            }

            public @Nullable String small = "";
            public @Nullable String medium = "";
            public @Nullable String large = "";
        }

        @Override
        public String toString() {
            return "Skill{" +
                    "id=" + id +
                    ", title='" + title + '\'' +
                    ", color='" + color + '\'' +
                    '}';
        }
    }

    class Challenge {
        public long id;
        public boolean active = true;
        public @Nullable String title;
        public @Nullable String description;
        public @Nullable Image image = new Image();

        public static class Image {
            public @Nullable Asset ios_600;
        }
    }

    class Asset {
        public @Nullable String url;
        public @Nullable String mime;
        public int width;
        public int height;
    }
}

File: app/src/main/java/com/bsstokes/bsdiy/api/DiyApi.kt
package com.bsstokes.bsdiy.api

import okhttp3.HttpUrl
import retrofit2.Response
import retrofit2.http.GET
import retrofit2.http.Path
import rx.Observable

interface DiyApi {

    companion object {
        const val ENDPOINT = "https://api.diy.org/"
    }

    @GET("/skills?limit=1000&offset=0")
    fun getSkills(): Observable<Response<DiyResponse<List<Skill>>>>

    @GET("/skills/{skill_url}/challenges")
    fun getChallenges(@Path("skill_url") skillUrl: String): Observable<Response<DiyResponse<List<Challenge>>>>

    object Helper {
        fun normalizeUrl(url: String): String {
            if (url.isBlank()) {
                return url
            }

            val httpUrl: HttpUrl? = if (url.startsWith("http", ignoreCase = true)) {
                HttpUrl.parse(url)
            } else {
                HttpUrl.parse("https:" + url)
            }

            return httpUrl?.toString() ?: ""
        }
    }

    data class DiyResponse<T>(var head: Head?, var response: T?) {
        data class Head(var code: Int?, var status: String?, var collection: Collection?) {
            data class Collection(var limit: Int?, var offset: Int?)
        }
    }

    data class Skill(var id: Long,
                     var active: Boolean?,
                     var url: String?,
                     var title: String?,
                     var description: String?,
                     var icons: Icons?,
                     var images: Images?,
                     var color: String?) {
        data class Icons(var small: String?, var medium: String?)
        data class Images(var small: String?, var medium: String?, var large: String?)
    }

    data class Challenge(var id: Long,
                         var active: Boolean?,
                         var title: String?,
                         var description: String?,
                         var image: Image?) {
        data class Image(var ios_600: Asset?)
    }

    data class Asset(var url: String?, var mime: String?, var width: Int?, var height: Int?)
}